raft/src/main/java/org/apache/kafka/raft/internals/BatchAccumulator.java

```diff
@@ -233,6 +235,68 @@ public void appendLeaderChangeMessage(LeaderChangeMessage leaderChangeMessage, l
         }
     }
 
+
+    public void appendSnapshotHeaderMessage(MetadataSnapshotHeaderRecord snapshotHeaderRecord) {
+        appendLock.lock();
+        try {
+            // Ideally there should be nothing in the batch here.
+            // TODO verify this
```

Comment:
I am concerned about improper usage in the future which might lead to stamping the header after other records in the file. Is asserting on the fact that the current batch has no records as this time valid? It should be an error if it isn't.

What do you think?

Reply:
It is a valid/correct check but it doesn't cover all of the cases. For example, in theory `SnapshotWriter` could do the following:
```
accumulator.appendSnapshotHeaderMessage(...);
snapshot.appendBatches(accumulator.drain());
accumulator.appendSnapshotHeaderMessage(...);
```

This is okay from the `BatchAccumulator`'s point of view because the accumulator doesn't have enough information. I am also not sure if we want to encode the invariant of `SnapshotHeaderMessage` and `SnaspshotFooterMessage` into the `BatchAccumulator`.

I think only the `SnapshotWriter` has enough information to make these invariant checks. What do you think?

Reply:
Hmmm. I see your point of not encoding the logic of maintaining the snapshot invariants into the BatchAccumulator, but then there is nothing in code preventing the usage pattern you listed above. Are we ok with that?

Reply:
I think that having tests that check this behavior for `SnaphsotWriter` is the best we can do.